File: solutions/ocaml/153-insert-delete-getrandom-o-1.ml
(*
  Problem 153: Insert Delete GetRandom O(1) (LeetCode 380)
  Difficulty: Med
  Language: OCaml
*)
type rand_set = { map: (int, int) Hashtbl.t; list: int array; size: int }

let create () = { map = Hashtbl.create 16; list = Array.make 128 0; size = 0 }

let insert rs v =
  if Hashtbl.mem rs.map v then false
  else begin
    Hashtbl.add rs.map v rs.size;
    rs.list.(rs.size) <- v;
    true
  end

let remove rs v =
  if not (Hashtbl.mem rs.map v) then false
  else begin
    let idx = Hashtbl.find rs.map v in
    let last_v = rs.list.(rs.size - 1) in
    rs.list.(idx) <- last_v;
    Hashtbl.replace rs.map last_v idx;
    Hashtbl.remove rs.map v;
    true
  end

File: solutions/rust/153-insert-delete-getrandom-o-1.rs
/**
 * Problem 153: Insert Delete GetRandom O(1) (LeetCode 380)
 * Difficulty: Med
 * Language: Rust
 */
use std::collections::HashMap;
use rand::Rng;
struct RandomizedSet { map: HashMap<i32, usize>, list: Vec<i32> }
impl RandomizedSet {
    fn new() -> Self { Self { map: HashMap::new(), list: vec![] } }
    fn insert(&mut self, val: i32) -> bool {
        if self.map.contains_key(&val) { return false; }
        self.map.insert(val, self.list.len());
        self.list.push(val); true
    }
    fn remove(&mut self, val: i32) -> bool {
        if let Some(idx) = self.map.remove(&val) {
            let last = *self.list.last().unwrap();
            self.list.swap_remove(idx);
            if idx < self.list.len() { self.map.insert(last, idx); }
            true
        } else { false }
    }
    fn get_random(&self) -> i32 {
        self.list[rand::thread_rng().gen_range(0..self.list.len())]
    }
}
